paywall/src/paywall-builder/web3Proxy.js

```diff
@@ -0,0 +1,88 @@
+import { enable } from './config'
+import { mainWindowPostOffice, setHandler } from '../utils/postOffice'
+import {
+  POST_MESSAGE_READY,
+  POST_MESSAGE_WEB3,
+  POST_MESSAGE_WALLET_INFO,
+} from './constants'
+
+let hasWeb3 = true
+
+/**
+ * Proxy calls to web3 from postMessage
+ *
+ * @param {window} window the main window object
+ * @param {element} iframe the iframe element, created by document.createElement('iframe')
+ * @param {string} origin the iframe element's URL origin
+ */
+export default function web3Proxy(window, iframe, origin) {
+  mainWindowPostOffice(window, iframe, origin)
+
+  // handler for the actual web3 calls
+  setHandler(POST_MESSAGE_WEB3, (payload, respond) => {
+    if (!hasWeb3) {
+      return respond(POST_MESSAGE_WEB3, {
+        id,
+        error: 'No web3 wallet is available',
+        result: null,
+      })
+    }
+
+    if (!payload || typeof payload !== 'object') return
+    if (!payload.method || typeof payload.method !== 'string') {
+      return
+    }
+    if (!payload.params || !Array.isArray(payload.params)) {
+      return
+    }
+    if (
+      typeof payload.id !== 'number' ||
+      Math.round(payload.id) !== payload.id
```

Comment:
Just curious, under what circumstances would `payload.id` not be an integer?

Reply:
if a rogue script (through XSS) was running on the main page trying to usurp control of the wallet